### PR TITLE
[SPARK-42892][SQL] Move sameType and relevant methods out of DataType

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

This PR moves the following methods from `DataType`:
1. equalsIgnoreNullability
2. sameType
3. equalsIgnoreCaseAndNullability

The moved methods are put together into a Util class.

### Why are the changes needed?

To make `DataType` become a simpler interface, non-public methods can be moved outside of the DataType class.

### Does this PR introduce _any_ user-facing change?

No as the moved methods are private within Spark.

### How was this patch tested?

Existing UT.

### CODE REVIEW

**Comment by amaliujia** [comment]
@hvanhovell @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/types/DataTypeUtils.scala [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    if (SQLConf.get.caseSensitiveAnalysis) {
+      equalsIgnoreNullability(left, right)
+    } else {
+      equalsIgnoreCaseAndNullability(left, right)
+    }
+
+  /**
+   * Compares two types, ignoring nullability of ArrayType, MapType, StructType.
+   */
+  def equalsIgnoreNullability(left: DataType, right: DataType): Boolean = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/types/DataTypeUtils.scala:37]: does this method require any non-public APIs?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/types/DataTypeUtils.scala:37]: nvm, it was a static method before this PR.

**Review by cloud-fan — APPROVED** [review]

**Review by hvanhovell — APPROVED** [review]
LGTM

**Pull request closed without merging**